kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -771,8 +770,7 @@ public class TypeSpec private constructor(
         "typevariables are forbidden on anonymous types"
       }
 
-      val isAbstract = ABSTRACT in modifiers || SEALED in modifiers || kind != Kind.CLASS ||
```

Comment:
`kind != Kind.CLASS` is the condition that is allowing objects to have abstract functions